llvm/lib/ObjCopy/ELF/ELFObject.cpp

```diff
@@ -2635,9 +2635,36 @@ template <class ELFT> Error ELFWriter<ELFT>::finalize() {
 }
 
 Error BinaryWriter::write() {
-  for (const SectionBase &Sec : Obj.allocSections())
+  SmallVector<const SectionBase *, 30> LoadableSections;
+  for (const SectionBase &Sec : Obj.allocSections()) {
+    if ((Sec.Flags & SectionFlag::SecLoad) && Sec.Type != SHT_NOBITS)
```

Comment:
Skipping NOBITS sections seems a fairly dramatic departure from the previous code. It may make sense, but I'd like to hear your explanation.

Reply:
This is interesting. The first part of this condition was wrong, I don't remember why I added it, so I removed it now. However, the NOBITS check should be there. NOBITS sections are not considered when assigning offsets and determining the size of the output buffer in `BinaryWriter::finalize()`. They are also not written to the output in `SectionWriter::visit()`. Another question is whether NOBITS sections should be filled with a pattern, but GNU objcopy does this, which makes sense to me.

Reply:
> Another question is whether NOBITS sections should be filled with a pattern, but GNU objcopy does this, which makes sense to me.

FWIW, the ELF spec explicitly states that `PT_LOAD` with `p_memsz > p_filesz` is padded with zeros (this is the case for a .bss section at a the end of a segment). It would be weird for other output formats to define a different system for initializing NOBITS data.

Anyway, no concrete concerns here, if @MaskRay is happy (like I said, I don't know much about the details of binary output).